### PR TITLE
Fix: Discard unfinished Spans before sending them over to Sentry

### LINKED ISSUE #1262: Discard unfinished Spans before sending them over to Sentry
We need to discard Spans that are unfinished before sending them over to Sentry, the reason is https://github.com/getsentry/develop/issues/274

Relay drops the whole transaction if a single Span is unfinished.

@Tyrrrz @brustolin I believe this is also important for you both

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :scroll: Description

Discard unfinished Spans before sending them over to Sentry

## :bulb: Motivation and Context

Fixes #1262

## :green_heart: How did you test it?

Unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by maciejwalkowiak** [comment]
A question from https://github.com/getsentry/sentry-java/issues/1262#issuecomment-782104801 must be answered before we can merge it

File: sentry/src/main/java/io/sentry/SentryClient.java [hunk 1]
```diff
@@ -409,6 +409,13 @@ public void captureSession(final @NotNull Session session, final @Nullable Objec
         }
       }
     }
+    final List<Span> unfinishedSpans = new ArrayList<>();
+    for (Span span : transaction.getSpans()) {
+      if (!span.isFinished()) {
+        unfinishedSpans.add(span);
+      }
+    }
+    transaction.getSpans().removeAll(unfinishedSpans);
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryClient.java:418]: Would be nice to log out. Maybe check if logging is enabled to warn level, if yes we log count of dropped span and the operations?

**Comment by marandaneto** [comment]
> A question from [#1262 (comment)](https://github.com/getsentry/sentry-java/issues/1262#issuecomment-782104801) must be answered before we can merge it

the answer is here: https://github.com/getsentry/develop/issues/274#issuecomment-782313995
let's do like python for now.

File: sentry/src/main/java/io/sentry/SentryClient.java (hunk 1 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryClient.java:418]: I'd log it too

**Review by marandaneto — APPROVED** [review]
I'd log unfinished spans and missing changelog, other than that, LGTM

**Review by bruno-garcia — COMMENTED** [review]
👍

File: sentry/src/main/java/io/sentry/SentryClient.java [hunk 2]
```diff
@@ -409,6 +409,18 @@ public void captureSession(final @NotNull Session session, final @Nullable Objec
         }
       }
     }
+    final List<Span> unfinishedSpans = new ArrayList<>();
+    for (Span span : transaction.getSpans()) {
+      if (!span.isFinished()) {
+        unfinishedSpans.add(span);
+      }
+    }
+    if (options.getLogger().isEnabled(SentryLevel.WARNING) && !unfinishedSpans.isEmpty()) {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryClient.java:418]:
```suggestion
    if (!unfinishedSpans.isEmpty()) {
```

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryClient.java:418]: `options.getLogger().log` already checks if its enabled internally, `!unfinishedSpans.isEmpty()` isn't a heavy call I believe, it'd be fine logging directly

Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/SentryClient.java:418]: 👍

**Pull request merged**